x-pack/plugins/security_solution/public/flyout/preview/hooks/use_rule_switch.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { useHasActionsPrivileges } from '../../../detection_engine/rule_management_ui/components/rules_table/use_has_actions_privileges';
+import { useHasMlPermissions } from '../../../detection_engine/rule_management_ui/components/rules_table/use_has_ml_permissions';
+import { isMlRule } from '../../../../common/machine_learning/helpers';
+import { useUserData } from '../../../detections/components/user_info';
+import {
+  canEditRuleWithActions,
+  explainLackOfPermission,
+  hasUserCRUDPermission,
+} from '../../../common/utils/privileges';
+import type { Rule } from '../../../detection_engine/rule_management/logic';
+export interface UseRuleSwitchParams {
+  /**
+   * Rule object that represents relevant information about a rule
+   */
+  rule: Rule | null;
+  /**
+   * Boolean that indicates whether the rule currently exists
+   */
+  isExistingRule: boolean;
+}
+
+export interface UseRuleSwitchResult {
+  /**
+   * Boolean of wheather useUserData is loading
+   */
+  userInfoLoading: boolean;
+  /**
+   * Tooltip text that explains why a user does not have permission to rule
+   */
+  tooltipText?: string;
+  /**
+   * Boolean that indivates whether the rule switch button is isabled
+   */
+  isButtonDisabled: boolean;
+  /**
+   * Boolean that indivates whether the rule switch shoud be enabled
+   */
+  isRuleEnabled: boolean;
+}
+
+/**
+ * This hook is used to retrieved data for rule switch
+ * @param rule data
+ * @param isExistingrule is this rule an existing rule
+ */
+export const useRuleSwitch = ({
```

Comment:
yeah I totally agree, this is my attempt of separating all the credential checks from the main component without refactoring all the rule switch code. It will be great if we can just import the button and use as is.

I'm torn because even if I moved `userInfoLoading` inside this hook, a lot of the boolean check are duplications of existing code on rules page. Down the line if any logics are changed, it's easy to neglect the rule switch in this flyout.

I'm actually tempted to drop the rule switch from this PR, as it was not part of the MVP criteria in the original ticket. I'll consult with the team and product, and keep you posted!